### PR TITLE
Add some advanced functionalities to ql module

### PR DESCRIPTION
**Pull request by yinan1048576**

### CODE REVIEW

**Comment by yinan1048576** [comment]
Add a rotate functionality.

**Comment by yinan1048576** [comment]
Add final couple of handy shorthand notations. Ready for review.

File: include/jemalloc/internal/ql.h [hunk 1]
```diff
@@ -58,6 +58,15 @@ struct {								\
 	ql_first(a_head) = qr_next((a_elm), a_field);			\
 } while (0)
 
+#define ql_concat(a_head_a, a_head_b, a_field) do {			\
+	if (ql_first(a_head_a) == NULL) {				\
+		ql_first(a_head_a) = ql_first(a_head_b);		\
+	} else if (ql_first(a_head_b) != NULL) {			\
+		qr_meld(ql_first(a_head_a), ql_first(a_head_b),		\
+		    a_field);						\
+	}								\
```
Comment (davidtgoldblatt) [inline include/jemalloc/internal/ql.h:74]: Let's null out a_head_b here? This makes this logically a "pop all" function. I think in most reasonable use cases the very next thing we'll want to do is reinitialize b to the empty list anyways (or never touch it again, in which case it doesn't end up mattering).

Comment (yinan1048576) [reply include/jemalloc/internal/ql.h:76]: Great point. Also added unit tests to verify.

**Comment by yinan1048576** [comment]
Rebase, and revised according to feedback. Added one more commit to better structure the `ql` module on top of the new utilities.

**Pull request merged**